ema.c

```diff
@@ -1198,17 +1198,28 @@ ema_t* ema_realloc_from_reserve_range(ema_t* first, ema_t* last, size_t start,
     }
 
     int ret = 0;
+    // Splitting nodes may add more emalloc reserve nodes.
+    // Those can be appended and move the "guard" which
+    // could be the last node
+    // We track the the last inclusive node.
+    ema_t* last_inclusive = last->prev;
     if (start > first->start_addr)
```

Comment:
Currently, this internal function `ema_realloc_from_reserve_range` is only used in the file sgx-mm.c, when the arguments `first`, `last`, `start` and `end` are determined by the function `search_ema_range`. And `search_ema_range` has the invariants  that `first->start_addr >= start` and `last->prev->start_addr + last->prev->size <=end`. Based on that, I think the checks were somewhat redundant at the first place. 
But if we regard `ema_realloc_from_reserve_range` as a separate function that does not require its arguments to meet above requirements, then we may need those checks and the fixes in it. But this way, that is regarding it as a separate function, we cannot ensure the assertions at the beginning are always true.

Reply:
The issue I saw was even though the assertions at the start of this function holds, the 'last' pointer can no longer be guaranteed valid 'last' when splitting happens and the split node from the last->prev requires a new emalloc reserve region to do the book keeping. Then the node for new reserve may get in between last->prev and last (e.g., if last was the root->guard).  If we don't keep last->prev, then the logic below would destroy all nodes from first to old last, which includes the new reserve node

Reply:
ah, I think there is a corner case not covered in ```search_ema_range``` in this loop.
```
   // find the last node that has addr <= 'end'
    while ((node != root->guard) && (!ema_higher_than_addr(node, end)))
    {
        node = node->next;
    }
    *ema_end = node;
```
 when node==root->guard && node->prev->start<end && last->prev->start+size>end. In this case, last will be guard and last->prev needs be split later in realloc.  We may need redefine ```ema_end``` to be the last node overlapping with given range and make it processed in all callers of   ```search_ema_range```

Reply:
I checked `search_ema_range`. The function returns the list of ema nodes that overlap the address range represented by `start` and `end` , so yes, the checks after `ema_split` is necessary.